### ISSUE TITLE
Using unit-formula

### ISSUE THREAD

**Issue by nneuss** [Ramarren/unit-formula#5]
Hello,
I would be interested how I could use unit-formula in the following situation:
I have a set of physical parameters with units and would like to calculate from those dimensionless parameters for inserting them in a numerical calculation.
Should I declare those parameters all with define-units and then define formulae involving those parameters?
Thanks, Nicolas

**Comment by elliottjohnson** [comment]
Nneuss, can you provide an example?  I think I understand your question, but I could be missing your point.

I tend to define-units based upon existing units.  For dimensionless units I typically define them in terms of UNITY if I'm going to be making a formula[e] for them.  For most dimensionless units, they can be defined purely in relation to a unit-less quantity.  For instance:

    (define-unit amphora (27.18 liters))          ; A typical unit definition
    (define-unit parts-per-million 0.000001)  ; A typical dimensionless unit 

In most of my use cases I've taken a collection of raw data (say numerical values that express quantities in liters) and used REDUCE-UNIT to convert the raw values into a unit that I can pass to a formula.

If your values are truly dimensionless and you'd like to convert them to a dimensioned unit, then you could do so with a formula like this:

    UNIT-FORMULAS> (defformula* to-gallons ((u unity)) (* u gallons))
    UNIT-FORMULAS> (to-gallons 1)
    #<UNIT 0.0037854117647058826d0 m^3>

Or you can do this without a formula, again using REDUCE-UNIT:

    UNIT-FORMULAS> (let ((my-dimensionless-unit 100))
    		 (reduce-unit `(* ,my-dimensionless-unit
    				  gallons)))
    #<UNIT 0.37854117647058827d0 m^3>
    UNIT-FORMULAS> 

Hopefully I'm understanding your question correctly.
-Elliott

**Comment by nneuss** [comment]
Let us say that I want to specify and solve a 1D-diffusion problem of the form
$$ \partial_t u - D \partial_{xx} u = f $$
Now, D=1.76e-5 m^2/s is a value from Wikipedia (oxygen in air).
1. How should I specify this parameter?  Something like this might be convenient:
   (defparameter *D_Oxygen*
     (constant-with-dimension '(1.76e-5 (/ (* m m) s))))
2. Next, the experimental setup has only a diameter of a few centimeters, so it is convenient that coordinates are measured in centimeters which could be specified as follows:
   (defparameter *my-length-unit* 'cm)
   (defparameter *my-time-unit* 's)
3. The simulation then needs internally to work out a diffusion coefficient measured in [cm^2/s] and has to do something like
   (let ((D (numerical-value-of
             (convert *D-oxygen*
                      `(/ (* ,*my-length-unit* ,*my-length-unit*)
                          ,*my-time-unit*)))))
     ;; calculation then uses this numerical value of D
     )
How could I do this conveniently with unit-formulas?

Thank you,
Nicolas

**Comment by nneuss** [comment]
3. And more complicated: Assume that D is specified by a formula like D=alpha*temperature where alpha is a tabular value measured in [m^2/(s K)].  How should I calculate D internally in this case?  I would need something like the following:

(defparameter *alpha* '(2e-3 (/ (* m m) s K)))
(defparameter *my-temperature* '(298 K))

(let ((D (numerical-value-of
          (convert (* ,*alpha, ,*my-temperature*)
                   `(/ (* ,*my-length-unit* ,*my-length-unit*)
                       ,*my-time-unit*)))))
  ;; calculation then uses this numerical value of D
  )

**Comment by elliottjohnson** [comment]
Nneuss,

You can specify the parameter in any form you choose, but I typically reduce the unit into its internal representation:

    (defmacro defconst-with-dimension (name value)
      `(defparameter ,name (reduce-unit ,value)))
    (defconst-with-dimension d_oxygen '(1.76e-5 (/ (* m m) s)))
    (defconst-with-dimension alpha '(2e-3 (/ (* m m) s K)))
    (defconst-with-dimension my-temperature '(298 K))

The result held in each parameter is a unit object with a factor value scaled to unit-formulas' base units (in this case the units you've supplied are base units, so there is no change in the factor).

When there is a need to plug these values into a formula, defformula is appropriate (I happen to like the defformula* version, but defformula does more bulletproofing).  Here is a trivial example:

    (defformula* fake-diffusion-time ((area (* m m))
                                      (coefficient (/ (* m m) s)))
		 (* area (/ 1 coefficient)))

which will automatically handle any units of input:

    UNIT-FORMULAS> (fake-diffusion-time '(2 (* ft ft)) d_oxygen)
    #<UNIT 10557.163636363637d0 s>

In your actual formula will you be using only the following operators or will you need to perform calculus?:

    IF < > <= >= + - ABS * / EXPT SQRT SIN COS TAN SINH COSH TANH ASIN ACOS ATAN
    ASINH ACOSH ATANH LOG EXP

If you have a formula that needs to do derivatives, then defformula[e*] unfortunately handle the math.  I believe the code can be expanded to handle it, but as it stands defformula is rather limited.

What I would suggest and what I typically do is to use unit-formulas for what it does well, which is unit conversion.  I typically write functions that accept unit-formula:units as inputs and inside the function I pull out the unit in the form I need it, do math on it and then return the value as a unit.  Here's some basic example code:

    UNIT-FORMULAS> (defgeneric my-formula (unit1 unit2)
                              (:method ((unit1 unit)(unit2 unit))
                                        (let ((length (convert-unit unit1 'meters))
                                                (time (convert-unit unit2 'seconds)))
                                          (reduce-unit (list (/ length time) '(/ meters second))))))
    #<STANDARD-GENERIC-FUNCTION UNIT-FORMULAS::MY-FORMULA (1)>
    UNIT-FORMULAS> (defmethod my-formula ((unit1 list)(unit2 list))
                                      (my-formula (reduce-unit unit1) (reduce-unit unit2)))
    #<STANDARD-METHOD UNIT-FORMULAS::MY-FORMULA (LIST LIST) {1006850913}> 
    UNIT-FORMULAS> (my-formula '(3 ft) '(1 hour))
    #<UNIT 2.5400000000000005d-4 m s^-1>

Typically I'll write a macro that builds methods that allow for all of the permutations of input types, including the use of some sensible defaults (ie. just a number provided as the first argument will be assumed in meters... yes I get lazy sometimes).  I think this will allow you to express the math you need with the flexibility of tossing random units at your equations and getting correct results.  I believe this would also allow you to calculate your diffusion coefficient as well (part #4).

Long story short, defformula is not that greatest feature of this library, but for simple (and repetitious) conversions and calculations it does do a great job.  

Hopefully this has been a help.

-Elliott

**Comment by nneuss** [comment]
Thank you Elliott,

I think, the combination of convert-unit and reduce-unit does most of what I want.  If I do the following:
`(defmacro defconst-with-dimension (name value)
  `(defparameter ,name ,value))

(defconst-with-dimension *D-oxygen* '(1.76e-5 (/ (* m m) s)))
(defconst-with-dimension *alpha* '(2e-3 (/ (* m m) s K)))
(defconst-with-dimension *my-temperature* '(298 K))

(defconst-with-dimension *my-length-unit* '(1 mm))
(defconst-with-dimension *my-time-unit* '(1 s))

(let ((D (convert-unit
          (reduce-unit *D-oxygen*)
          `(/ (* ,*my-length-unit* ,*my-length-unit*)
              ,*my-time-unit*))))
  D  ; use this value
  )

(let ((D (convert-unit
          (reduce-unit `(* ,*alpha* ,*my-temperature*))
          `(/ (* ,*my-length-unit* ,*my-length-unit*)
              ,*my-time-unit*))))
  D  ; use this value
)
`
If this is reasonable, then probably a combined function with two parameters would be appropriate, as well as some checks in defconst-with-dimension (note that I had to drop the reduce-unit there, because otherwise the calculation later on would have had to be done in a more complicated way with defformula).

Thanks again,
Nicolas

**Comment by nneuss** [comment]
Hmm, the code formatting does not work in combination with Lisp's quasiquote notation:-(

**Comment by elliottjohnson** [comment]
Nicolas,

You can use unit-formulas in this way.  There isn't a lot gained by using the library in this method, but it can be done.  The real power of unit-formulas is passing UNIT objects into functions.  This way you can convert the UNIT to the units needed in a formula, then perform the math, and then return the value as a UNIT, which can be later converted into any unit needed by the user.

The advantage is that a user can supply you with numbers in any units that fulfill the required dimensions (ie. area, time, and temperature) and the function will return a UNIT that then can be easily converted into any unit they may need the answer in.  In some professions (such as my position as a winemaker in the U.S.) there is a need to go between many different units at any given time (gallons <=> liters, feet & inches <=> mm, ppm <=> mg/L *yes this is a rough approximation) and this usage pattern is very effective.

Looking at the forms you are working with, I would suggest:

    (defpackage :diffusion
      (:use :cl :unit-formulas))
    
    (in-package :diffusion)
    
    (defmacro defconst-with-dimension (name value)
      `(defparameter ,name (reduce-unit ',value)))
    
    (defconst-with-dimension D-oxygen (1.76e-5 (/ (* m m) s)))
    (defconst-with-dimension alpha (2e-3 (/ (* m m) s K)))
    (defconst-with-dimension my-temperature (298 K))

    (defgeneric diffusion-area (time &optional temperature)
      (:method ((time unit) &optional (temperature my-temperature))
        (assert (or (null temperature)
    		(listp temperature)
    		(typep temperature 'unit)))
        (let ((d (if temperature
    		 (reduce-unit (list '* alpha temperature))
    		 d-oxygen)))
          (reduce-unit (list '* time d))))
      (:method ((time list) &optional (temperature my-temperature))
        (diffusion-area (reduce-unit time) temperature))
      (:method ((time number) &optional (temperature my-temperature))
        (diffusion-area (reduce-unit (list time 'seconds)) temperature)))

In the first method used above you can substitute any math you need to use within the body of the LET.  Once you are ready to return a value the REDUCE-UNIT form can be used to return a UNIT with the appropriate units.  Some usage examples are:

    DIFFUSION> (diffusion-area '(1 sec))
    #<UNIT 0.596d0 m^2>
    DIFFUSION> (diffusion-area '(1 sec) '(1 K))
    #<UNIT 0.002d0 m^2>
    DIFFUSION> (diffusion-area 100)
    #<UNIT 59.599999999999994d0 m^2>
    DIFFUSION> (diffusion-area '(100 seconds))
    #<UNIT 59.599999999999994d0 m^2>
    DIFFUSION> (diffusion-area '(100 min) '(400 K))
    #<UNIT 4800.0d0 m^2>

The importance of returning a UNIT object is that it then can be converted into any required unit, which is especially important for countries that use antiquated units of measure:

    DIFFUSION> (convert-unit * '(* mile mile))
    2.301168864912977d-7

Hopefully this gives you some ideas about how to use this library and how to frame your calculations.  There are a lot of options to explore.  Let me know if you have any other questions...


Regards,
-Elliott

PS - I use four spaces at the start of each line, which allows for quasiquoting to show correctly.  I hope that is useful in future interactions with github.

**Comment by nneuss** [comment]
Hi Elliott,
thank you for your answer.
The problem is that I cannot work with units throughout my PDE library (Femlisp), because I would then have to store mesh coordinates etc. all with units.  This is only a theoretical possibility because it would slow down calculations enormously.  Thus, I need some interface below(?) which only calculations with numbers (more precisely: only floating point numbers) are done.
For the moment, the least intrusive way to introduce units is probably to calculate the numerical values of some parameters.  Another a bit more far-reaching idea would be something where a fast function is generated from a function working with units:

`    (defconst-with-dimension alpha (2e-3 (/ (* m m) s K)))

    ;;; formula resulting in a function Diffusion working with units
    (define-formula Diffusion (Temperature)
      (* alpha Temperature))

    ;; Returns a function which works with numbers.  It expects a numeric
    ;; argument meaning temperature in K and returns a number meaning diffusion
    ;; in the unit (/ (* mm mm) s).

    (instantiate-formula Diffusion '(K) '(/ (* mm mm) s))
    ;; the result being:
    => (lambda (Temperature) (* 2e-3 1000000 Temperature))
    == (lambda (Temperature) (* 2e3 Temperature))
`

**Comment by elliottjohnson** [comment]
Nicolas,

I would think you would store unit data on a mesh basis, since as you mention it doesn't make much sense to store per mesh-item units.  I could even see the use of a mesh specialization with the currently defined methods.  Abstracting functionality, I could see a method used to convert grids from one unit to another making meshes a first class, unit-aware, object.  Maybe this is not the best approach, since it ties more into my interest in unit-formulas being a units focused library rather than formula focused, which was the original intention of Ramarren.

Congratulations on FEMLisp.  Quite a library and bigger than any open source project I've yet worked on.  Quite an accomplishment on putting that together based upon your research and publication.  Let me know if there is anything more I can do to help.

Regards,
Elliott

**Comment by nneuss** [comment]
Hi Elliott,
working with defparamete is not good at this point, because asterisks around units would be ugly and special variables without asterisks are problematic.  Thus, I do the following now which works quite nice up to now:
`
    (defmacro register-unit (name value)
      `(progn (define-units ,name ,value)
              (format t "~20A = ~A~%" ',name (reduce-unit ',name))))

    (register-unit Temp (298 K))
    (register-unit eta (1e-3 (/ (* N s) (* m m))))
    (register-unit R (22e-9 m))
    (register-unit Rh (60e-9 m))
    (register-unit D (/ (* boltzmann Temp) (* 6 pi eta Rh)))
    ;;; and many more ...

    ;;; and when setting up the problem
    (defun setup-problem
        (&key
         (D (convert-unit 'D '(/ (* mm mm) s)))
           ...)
      ...)
`
We already found a serious bug using this setup! Very nice:-)

So thanks again for your help,
Nicolas
